src/main/java/org/osc/manager/ism/api/IsmDeviceApi.java

```diff
@@ -138,80 +189,217 @@ public Void call() throws Exception {
+                } else {
+                    query.select(r).where(
+                            criteriaBuilder.and(criteriaBuilder.equal(r.get("vsId"), vsId)));
+                }
+
+                List<DeviceEntity> devices = IsmDeviceApi.this.em.createQuery(query).getResultList();
+                if (devices.isEmpty()) {
+
+                    String msg = String.format("Cannot find the device for " + "id: %s, VS id: %s\n",
+                            entity.getParent().getId(), IsmDeviceApi.this.vs.getId());
+                    LOG.error(msg);
+                    throw new IllegalArgumentException(msg);
+                }
+                CriteriaBuilder memberBuilder = IsmDeviceApi.this.em.getCriteriaBuilder();
+                CriteriaQuery<DeviceMemberEntity> memberQuery = memberBuilder.createQuery(DeviceMemberEntity.class);
+                Root<DeviceMemberEntity> res = memberQuery.from(DeviceMemberEntity.class);
+                memberQuery.select(res)
+                .where(memberBuilder.and((memberBuilder.equal(r.get("name"), entity.getName()))));
+                List<DeviceMemberEntity> deviceMemberResult = IsmDeviceApi.this.em.createQuery(memberQuery)
+                        .getResultList();
+                if (!deviceMemberResult.isEmpty()) {
+                    String msg = String.format("Device member already exists " + "name: %s\n", entity.getName());
+                    LOG.error(msg);
+                    throw new IllegalArgumentException(msg);
+                }
+                entity.setParent(devices.get(0));
+                IsmDeviceApi.this.em.persist(entity);
+                return entity;
             }
-        }).getId().toString();
+        }).getId();
     }
 
 
     @Override
     public String updateDeviceMember(ManagerDeviceMemberElement deviceElement, String name, String vserverIpAddress,
             String contactIpAddress, String ipAddress, String gateway, String prefixLength) throws Exception {
-        return deviceElement.getId();
+
+        DeviceMemberEntity deviceMember = new DeviceMemberEntity();
+        deviceMember.setId(Long.parseLong(deviceElement.getId()));
+        deviceMember.setName(name);
+        return updateDeviceMember(deviceMember, null, this.vs.getId());
+        // TODO - Sudhir Add other information like gateway, ipAddress
+    }
+
+    public String updateDeviceMember(DeviceMemberEntity deviceElement, Long parentId, Long vsId) throws Exception {
+
+        return this.txControl.required(new Callable<DeviceMemberEntity>() {
+
+            @Override
+            public DeviceMemberEntity call() throws Exception {
+
+                CriteriaBuilder criteriaBuilder = IsmDeviceApi.this.em.getCriteriaBuilder();
+
+                CriteriaQuery<DeviceMemberEntity> query = criteriaBuilder.createQuery(DeviceMemberEntity.class);
+                Root<DeviceMemberEntity> r = query.from(DeviceMemberEntity.class);
```

Comment:
This code to validate the existence of the parent is being used in multiple places. Make it a private common method and reuse instead of repeating.